File: leetcode/tree/589.n叉树的前序遍历.js
/*
 * @lc app=leetcode.cn id=589 lang=javascript
 *
 * [589] N叉树的前序遍历
 *
 * https://leetcode-cn.com/problems/n-ary-tree-preorder-traversal/description/
 *
 * algorithms
 * Easy (74.05%)
 * Likes:    111
 * Dislikes: 0
 * Total Accepted:    50.8K
 * Total Submissions: 68.7K
 * Testcase Example:  '[1,null,3,2,4,null,5,6]'
 *
 * 给定一个 N 叉树，返回其节点值的前序遍历。
 *
 * 例如，给定一个 3叉树 :
 *
 *
 *
 *
 *
 *
 *
 * 返回其前序遍历: [1,3,5,6,2,4]。
 *
 *
 *
 * 说明: 递归法很简单，你可以使用迭代法完成此题吗?
 */

// @lc code=start
/**
 * // Definition for a Node.
 * function Node(val, children) {
 *    this.val = val;
 *    this.children = children;
 * };
 */

/**
 * @param {Node} root
 * @return {number[]}
 */
// recursive
var preorder = function (root) {
  const ret = [];
  dfs(root);
  return ret;

  function dfs(root) {
    if (root) {
      ret.push(root.val);
      if (root.children?.length) {
        for (const child of root.children) {
          dfs(child);
        }
      }
    }
  }
};

// iterative
var preorder = function (root) {
  const ret = [];
  if (!root) return ret;
  const stack = [];
  stack.push(root);

  while (stack.length) {
    const node = stack.pop();
    ret.push(node.val);
    let len = node.children.length;
    if (len) {
      for (let i = len - 1; i >= 0; i--) {
        stack.push(node.children[i]);
      }
    }
  }

  return ret;
};
// @lc code=end

File: leetcode/tree/589.n-叉树的前序遍历.go
/*
 * @lc app=leetcode.cn id=589 lang=golang
 *
 * [589] N 叉树的前序遍历
 *
 * https://leetcode-cn.com/problems/n-ary-tree-preorder-traversal/description/
 *
 * algorithms
 * Easy (75.93%)
 * Likes:    261
 * Dislikes: 0
 * Total Accepted:    136.8K
 * Total Submissions: 180.2K
 * Testcase Example:  '[1,null,3,2,4,null,5,6]'
 *
 * 给定一个 n 叉树的根节点  root ，返回 其节点值的 前序遍历 。
 *
 * n 叉树 在输入中按层序遍历进行序列化表示，每组子节点由空值 null 分隔（请参见示例）。
 *
 *
 * 示例 1：
 *
 *
 *
 *
 * 输入：root = [1,null,3,2,4,null,5,6]
 * 输出：[1,3,5,6,2,4]
 *
 *
 * 示例 2：
 *
 *
 *
 *
 * 输入：root =
 * [1,null,2,3,4,5,null,null,6,7,null,8,null,9,10,null,null,11,null,12,null,13,null,null,14]
 * 输出：[1,2,3,6,7,11,14,4,8,12,5,9,13,10]
 *
 *
 *
 *
 * 提示：
 *
 *
 * 节点总数在范围 [0, 10^4]内
 * 0 <= Node.val <= 10^4
 * n 叉树的高度小于或等于 1000
 *
 *
 *
 *
 * 进阶：递归法很简单，你可以使用迭代法完成此题吗?
 *
 */

package leetcode

// Definition for a Node.
type Node struct {
	Val      int
	Children []*Node
}

// @lc code=start
func preorder2(root *Node) []int {
	var res []int
	var dfs func(root *Node)
	dfs = func(root *Node) {
		if root == nil {
			return
		}
		res = append(res, root.Val)
		for _, child := range root.Children {
			dfs(child)
		}
	}
	dfs(root)
	return res
}

func preorder(root *Node) []int {
	if root == nil {
		return nil
	}
	var res []int
	var stack []*Node
	stack = append(stack, root)
	for len(stack) > 0 {
		node := stack[len(stack)-1]
		stack = stack[:len(stack)-1]
		res = append(res, node.Val)
		for i := len(node.Children) - 1; i >= 0; i-- {
			stack = append(stack, node.Children[i])
		}
	}
	return res
}

// @lc code=end
